File: Day171_2025/Q171_Leetcode.py
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        latitude = 0
        longitude = 0
        ans = 0
        n = len(s)

        for i in range(n):
            if s[i] == 'N':
                latitude += 1
            elif s[i] == 'S':
                latitude -= 1
            elif s[i] == 'E':
                longitude += 1
            elif s[i] == 'W':
                longitude -= 1

            manhattan = abs(latitude) + abs(longitude)
            ans = max(ans, min(manhattan + 2 * k, i + 1))

        return ans

File: Day171_2025/Q171_Leetcode.java
class Q171_Leetcode {
    public int maxDistance(String s, int k) {
        int latitude = 0, longitude = 0, ans = 0;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            char dir = s.charAt(i);
            if (dir == 'N') latitude++;
            else if (dir == 'S') latitude--;
            else if (dir == 'E') longitude++;
            else if (dir == 'W') longitude--;

            int manhattan = Math.abs(latitude) + Math.abs(longitude);
            ans = Math.max(ans, Math.min(manhattan + k * 2, i + 1));
        }

        return ans;
    }
}
